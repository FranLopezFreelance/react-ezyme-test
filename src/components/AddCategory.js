import React, { useState } from 'react';
import PropTypes from 'prop-types';

const AddCategory = ({setCategory}) => {
  
  const [inputValue, setInputValue] = useState('');

  const handleInputChange = (e) => {
    setInputValue(e.target.value)
  }

  const handleAddCategory = () => {
    if(inputValue.trim().length > 1){
      setCategory(cats => [inputValue, ...cats ]);
      setInputValue('');
    }
  }

  const handleSubmit = (e) => {
    e.preventDefault();
    if(inputValue.trim().length > 1){
      setCategory(cats => [inputValue, ...cats ]);
      setInputValue('');
    }
  }

  return (
    <form onSubmit={handleSubmit}>
      <input 
        placeholder="Buscar... "
        type="text" 
        className="category-input" 
        value={inputValue} 
        onChange={handleInputChange} />
      <button 
        className="add-category"
        onClick={handleAddCategory}>Buscar</button>
    </form>
  )
}

AddCategory.propTypes = {
  setCategory: PropTypes.func.isRequired
}

export default AddCategory
