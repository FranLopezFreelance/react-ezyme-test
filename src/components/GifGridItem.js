import React from 'react'
import PropTypes from 'prop-types';

export const GifGridItem = ({url, title}) => {
  return (
    <div className="gif">
      <p>{title}</p>
      <img className="gif-img" src={url} alt={title} />
    </div>
  )
}

GifGridItem.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired
}

export default GifGridItem;
