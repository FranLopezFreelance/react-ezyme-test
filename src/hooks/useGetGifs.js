import { useState, useEffect } from "react"
import { getGifs } from '../helpers/getGifs';

export const useGetGifs = (category) => {

  const [state, setState] = useState ({
    images: [],
    loading: true
  });

  useEffect(() => {
    getGifs(category).then(images => {
      setTimeout(() => {
        setState({
          images,
          loading: false
        });
      }, 4000);
    });
  }, [category]);

  return state;
}
