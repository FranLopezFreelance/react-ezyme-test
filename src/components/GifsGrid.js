import React, { useState, useEffect, Fragment } from 'react';
import PropTypes from 'prop-types';
import { GifGridItem } from './GifGridItem';

export const GifsGrid = ({category}) => {

  const [images, setImages] = useState([]);
  useEffect(() => {
    getGifs();
  }, []);

  const apiKey = 'eU0RKJE15pCeTK4baaf03vzbKUMso0x3';
  const url = `https://api.giphy.com/v1/gifs/search?q=${encodeURI(category)}&limit=12&api_key=${apiKey}`;
  const getGifs = async () => {
    const res = await fetch(url);
    const {data} = await res.json();
    let gifs = [];
    if(data){
      gifs = data.map(({id, title, images}) => {
        const {downsized_medium: url} = images;
        return {id, title, url: url.url};
      });
    }
    
    setImages(gifs);
  }

  return (
    <Fragment>
      <h3>{category}</h3>
        {
          images.map(image => {
            return <GifGridItem 
                      key={image.id} 
                      {...image} />
          })
        }
    </Fragment>
  )
}

GifsGrid.propTypes = {
  category: PropTypes.string.isRequired
}

export default GifsGrid;
