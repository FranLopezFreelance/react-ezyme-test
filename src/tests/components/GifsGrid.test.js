import React from 'react';
import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import { GifsGrid } from '../../components/GifsGrid';
import { useGetGifs } from '../../hooks/useGetGifs';

jest.mock('../../hooks/useGetGifs');

describe('Pruebas en el <GifsGrid />', () => {

  const category = 'Simpsons';

  test('Debe de mostrarse correctamente', () => {
    useGetGifs.mockReturnValue({
      images: [],
      loading: true
    });
    const wrapper = shallow(<GifsGrid category={category} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('Debe de mostrar items cuando se cargan imÃ¡genes con useGetGifs', () => {
    const images = [
      {
        id: 'ABC',
        url: 'http://url.com/imgs/1.jpg',
        title: 'Cualquier cosa'
      },
      {
        id: 'BCD',
        url: 'http://url.com/imgs/1.jpg',
        title: 'Cualquier cosa'
      }
    ]
    
    useGetGifs.mockReturnValue({
      images,
      loading: false
    });
    const wrapper = shallow(<GifsGrid category={category} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('p').exists()).toBe(false);
    expect(wrapper.find('GifGridItem').length).toBe(images.length);
  });
  
});
