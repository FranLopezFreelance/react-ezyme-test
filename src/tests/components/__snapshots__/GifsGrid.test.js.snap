// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <GifsGrid /> Debe de mostrar items cuando se cargan im√°genes con useGetGifs 1`] = `
<div
  className="gif-item"
>
  <h4
    className="animate__animated animate__backInDown"
  >
    Simpsons
  </h4>
  <GifGridItem
    id="ABC"
    key="ABC"
    title="Cualquier cosa"
    url="http://url.com/imgs/1.jpg"
  />
  <GifGridItem
    id="BCD"
    key="BCD"
    title="Cualquier cosa"
    url="http://url.com/imgs/1.jpg"
  />
</div>
`;

exports[`Pruebas en el <GifsGrid /> Debe de mostrarse correctamente 1`] = `
<div
  className="gif-item"
>
  <h4
    className="animate__animated animate__backInDown"
  >
    Simpsons
  </h4>
  <p
    className="animate__animated animate__flash loading"
  >
    Buscando...
  </p>
</div>
`;
